src/impl-ecl.lisp

```diff
@@ -106,6 +106,26 @@ Distributed under the MIT license (see LICENSE file)
 (defun thread-yield ()
   (mp:process-yield))
 
+;;; Resource contention: semaphores
+
+;;; ECL prior to 16.2.0 has signal-semaphore bug causing only one thread wake up
+;;; at first call disregarding signal count. We depend on generic implementation
+;;; for these versions.
+#.(when (> ext:+ecl-version-number+ 160103)
+    `(progn
+       (defun make-semaphore (&key name (count 0))
+         (mp:make-semaphore :name name :count count))
+
+       (defun signal-semaphore (semaphore &optional (count 1))
+         (mp:signal-semaphore semaphore count))
+
+       (defun wait-on-semaphore (semaphore &key timeout)
+         (if timeout
+             (handler-case (with-timeout (timeout)
```

Comment:
Is the use of WITH-TIMEOUT safe ?

Reply:
It won't crash nor anything. ECL has some limitation for timeouts – you can't interrupt sleeping process. So 

In sense of crashing it is safe (there should be nothing wrong with it). It has a drawback, because nesting timeouts doesn't work (you can't interrupt process which is during the call to SLEEP like that). Effectively if you do:

```
(with-timeout (1) (wait-on-semaphore … :timeout 2))
```

it will end after two seconds. Compare to much simpler:

```
CL-USER> (time (handler-case (bt:with-timeout (0.1) (sleep 1) "hi")
                   (bt:timeout () 3)))
real time : 1.001 secs
run time  : 0.000 secs
gc count  : 1 times
consed    : 5395424 bytes
```

That said, I want to add different kind of timeout (interruptible) to both semaphores and condition-variables, but I don't know if I will manage to do it before ECL 16.2.0 release.